### DISCUSSION TITLE
Release Testing 2023.05.1 - RC1

### DISCUSSION THREAD

**Discussion by tedpatrick** [pyscript/pyscript#1489] [Announcements]
We are feature complete for a release of PyScript next week. We need your help testing 2023.05.1.

How to test 2023.05.1.RC1

```
<link rel="stylesheet" href="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.css"/>
<script defer src="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.js"></script>
```

Thank you for your ongoing support and contribution to PyScript.

2023.05.1.RC1 Example:

```
<html>
<head>
    <link rel="stylesheet" href="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.css" />
    <script defer src="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.js"></script>
</head>
<body>
    <py-config type="toml">
packages = [
    "pandas"
]
    </py-config>
    <py-script>
import js
import pandas as pd

js.console.log( "2023.05.1.RC1" )
js.document.title = "2023.05.1.RC1"
print( "2023.05.1.RC1" )
print( pd.show_versions() )
    </py-script>
</body>
</html>
```

**Comment by bugzpodder** [comment]
tried pyscript.min.js, lgtm

**Comment by MovingBitsGroupRoth** [comment]
Works perfectly well on my side:

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>PyScript</title>
    <link rel="stylesheet" href="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.css"/>
    <script defer src="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.js"></script>
</head>
<body>

<div>Viper max: <span id="manual-write"></span></div>


<py-config type="toml">
    packages = [
        "pandas"
    ]
</py-config>

<py-script>
    import js
    import pandas as pd

    js.console.log( "Console: 2023.05.1.RC1" )
    js.document.title = "The PyScript version: 2023.05.1.RC1"
    print( "Volla says: 2023.05.1.RC1" )

    sep = "-" * 40

    print( pd.show_versions() )
    print(sep)

    df = pd.DataFrame([[1, 2], [4, 5], [7, 8]],
    index=['cobra', 'viper', 'sidewinder'],
    columns=['max_speed', 'shield'])
    print(df)
    print(sep)

    print(df.iloc[2])
    print(sep)

    viper_max_speed = df.loc['viper', 'max_speed']

    manual_div = Element("manual-write")
    manual_div.element.innerHTML = viper_max_speed

</py-script>
</body>
</html>
```

**Comment by bleach86** [comment]
py-* attributes that execute async python functions are no longer awaited.

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>PyScript</title>
    <link rel="stylesheet" href="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.css"/>
    <script defer src="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.js"></script>
</head>
<body>

<button id="py-button-to-click" py-click="hello()" type="button">Click Me!</button>

<py-script>
    import js
    import asyncio
    
    async def hello():
         js.console.log("Hello")

</py-script>
</body>
</html>
```

Running the above and clicking the button will result in the following error.

```

sys:1: RuntimeWarning: coroutine 'hello' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

Adding await to the call results in a await outside of async function error.

The solution that I have found to work is to wrap the function call in asyncio.create_task(), like so.

```html
<button id="py-button-to-click" py-click="asyncio.create_task(hello())" type="button">Click Me!</button>
```

This results in the expected execution of the async hello function.

**Reply by JeffersGlass** [reply]
I think the behvaior is the previous release was an an accident of implementation, and that the behavior in 230.05.1.RC1 is correct. 

Previously, the py-* event handlers were [handled by interpreter.run](https://github.com/pyscript/pyscript/blob/13e9252260c0b2c43c77e71deb4a5f0d0cada485/pyscriptjs/src/components/pyscript.ts#LL206C25-L206C60) which called `interface.runPython` directly. So if the user's code returned a Future or Coroutine, it would be scheduled because, if you resolve a Promise that wraps a thenable, [that thenable is immediately resolved/called](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve). (Thank you @hoodmane for that lesson in Lisbon).

Now after the reshaping, we redound to to `pyscript_internal.run_pyscript(code, id);` which avoids the 'awaiting a Promise of a thenable' issue.

So I think this is a change in behavior, and one we should document in the changelog (possibly as a bugfix?) Clearly at least one person was using the previous behavior in the wild.

**Comment by bleach86** [comment]
pyscript.runtime.globals.get('foo') syntax no longer works to execute python code from the javascript enviroment

```html

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>PyScript</title>
    <link rel="stylesheet" href="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.css"/>
    <script defer src="https://pyscript.net/snapshots/2023.05.1.RC1/pyscript.js"></script>
</head>
<body>

<py-script>
    async def foo(data):
        print("Data")
</py-script>
    <button onclick="pyscript.runtime.globals.get('foo')('Hello, world!')">Hello world?</button>
</body>
</html>
```

running the previous code in the RC results in the following error.

```
caught TypeError: pyscript.runtime.globals.get(...) is not a function
    at HTMLButtonElement.onclick (testfun.html:15:111)
```

In the previous versions this resulted in expected execution of the python function.

**Reply by madhur-tandon** [reply]
The internals have changed a bit due to ongoing support for workers...
You will have to add `await` statements, so the following should also work:
```
(await (await pyscript.runtime.globals.get('foo'))("Hello"))
```

**Reply by bleach86** [reply]
> Alternatively, you will have to add `await` statements, so the following should also work:
> 
> ```
> (await (await pyscript.runtime.globals.get('foo'))("Hello"))
> ```

Thanks. Using the 2 awaits does seem to get it working again.

**Reply by hoodmane** [reply]
I think it would be good to fix this so there is no backwards compatibility issue when `execution_mode` is `main_thread` and add a different API like `pyscript.asyncRuntime` for the worker. The synchronous APIs are actually useful and there is no technical reason we need to break them.

**Reply by JeffersGlass** [reply]
I agree with @hoodmane here, it would be nice if we could keep the synchronous API's the same. It'll prevent breaking a useful feature that I know is being used in the wild already. Especially since it's not entirely obvious how to stack the awaits to make the remote access work (at least not to me.)

**Reply by madhur-tandon** [reply]
https://github.com/pyscript/pyscript/pull/1529 should give back previous behaviour

**Comment by WebReflection** [comment]
@tedpatrick I would like to have @JeffersGlass [MR](https://github.com/pyscript/pyscript/pull/1317) part of the release, if Jeff is confident in merging it, as the work done in there is pretty awesome and it feels that it could sleep for long just for a couple of days / delay.

Thanks for considering this, I'd be OK if it's too late too though, hoping next release with that in would be sooner than later.

**Comment by tedpatrick** [comment]
PyScript 2023.05.1.RC2 https://github.com/pyscript/pyscript/discussions/1498
